Cache Spotify tokens per cookie and auto-refresh only anonymous token

Vexify4103 opened the pull request:
This PR updates the Spotify token handling to support per-cookie caching and proper refresh behavior:

- Introduces a token cache keyed by cookie sets (`cookiesKey`) to allow custom login cookies.
- Anonymous tokens (no cookies) are auto-refreshed in the background.
- Account tokens (with cookies) are cached but not automatically refreshed; they are fetched on-demand.
- Preserves previous fallback expiry logic in case the token response is malformed.
- Improves multi-user support by allowing simultaneous caching of different login sessions.
- Minor logging enhancements to indicate which token (key) is being returned or refreshed.

These changes ensure proper token management for both anonymous and logged-in Spotify users while avoiding unnecessary refreshes of account tokens.

Vexify4103 committed:
Add caching for Spotify token in handleToken

Introduces in-memory caching for the Spotify token response in handleToken, reducing redundant requests. The cache uses a mutex for thread safety and sets expiry based on the token's expiration timestamp if available, otherwise defaults to 55 minutes.

Vexify4103 committed:
Add background Spotify token refresher

Introduces a background goroutine to automatically refresh the Spotify token one minute before expiry. Refactors expiry parsing to use JSON unmarshalling for improved reliability. Updates main.go to start the refresher on server startup.

Vexify4103 committed:
Cache Spotify tokens by cookies; refresh only anonymous token

- Implement per-cookie token caching to support both anonymous and logged-in users.
- Anonymous token is auto-refreshed in the background.
- Account tokens (with cookies) are cached but not auto-refreshed; refreshed only on request.
- Introduce `cookiesKey()` helper to uniquely identify token cache entries per cookie set.
- Preserve previous fallback expiry logic for robustness.
- Improves handling of multiple simultaneous users and supports custom login cookies.

topi314 commented on handler.go:
```diff
@@ -2,16 +2,34 @@ package main
+var (
+	tokenCache = make(map[string]cachedTokenEntry)
+	cacheMutex sync.Mutex
+)
```
global state is bad

topi314 commented on handler.go:
```diff
@@ -2,16 +2,34 @@ package main
+)
+
+type spotifyTokenResponse struct {
+	AccessTokenExpirationMillis int64 `json:"accessTokenExpirationTimestampMs,string"`
```
field should be named the same as the json tag

topi314 commented on handler.go:
```diff
@@ -24,6 +42,22 @@ func (s *server) handleToken(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write(entry.Token)
+		return
+	}
+	cacheMutex.Unlock()
```
locking and unlocking here adds a bug where the token could be fetched multiple times if 2 requests happen basically at the same time

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+
+
+func parseExpiry(body []byte) time.Time {
+	expiry := time.Now().Add(55 * time.Minute) // fallback default
```
why default to 55 minutes?
pretty sure its 60 minutes atm

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+	go func() {
+		for {
+			cacheMutex.Lock()
+			entry, exists := tokenCache["anonymous"]
```
I'd rather use `""` for anonymous since that request essentially has no cookies

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+
+func (s *server) startTokenRefresher() {
+	go func() {
+		for {
```
just use a `time.Ticker`
no time.Sleep needed

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
 }
+
+func (s *server) startTokenRefresher() {
+	go func() {
```
move the go into the main function

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+				continue
+			}
+
+			sleepDuration := time.Until(entry.Expiry.Add(-1 * time.Minute))
```
how does spotify work here?
if I request a token 1min before expiry, does it return the in theory still valid token or a new one?

Vexify4103 replied:
I don't know how this works here.

Vexify4103 replied:
fixed in recent push, by changing auto refresh to 100ms after expiry

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+	expiry := time.Now().Add(55 * time.Minute) // fallback default
+
+	var resp spotifyTokenResponse
+	if err := json.Unmarshal(body, &resp); err == nil && resp.AccessTokenExpirationMillis > 0 {
```
idk, if we don't get json back we should probably fail here

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+
+func cookiesKey(cookies []*network.CookieParam) string {
+	if len(cookies) == 0 {
+		return "anonymous"
```
I'd rather use an empty string here

topi314 commented on handler.go:
```diff
@@ -100,3 +146,69 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+	}
+	var parts []string
+	for _, c := range cookies {
+		parts = append(parts, c.Name+"="+c.Value)
```
you can just call `c.String()`

topi314 replied:
> Applied most requested changes. c.String() does not exist on c..
> 
> > network.CookieParam has no String() method

just don't iterate over the cookies a second time for no reason

Vexify4103 committed:
fix: apply requested handler.go improvements

- Remove global tokenCache and cacheMutex; moved into server struct
- Rename spotifyTokenResponse field to match JSON tag
- Fix cache locking: prevent multiple simultaneous token fetches
- Use 60-minute default expiry instead of 55 minutes
- Use "" as key for anonymous requests (no cookies)
- Replace manual sleep loops with time.Ticker in token refresher
- Move goroutine start to main function
- Update parseExpiry to fail if JSON token missing
- Update cookiesKey to return "" instead of "anonymous"
- cookiesKey: cannot use c.String() as suggested; *network.CookieParam has no String() method

Vexify4103 commented:
Applied most requested changes.
c.String() does not exist on c..
> network.CookieParam has no String() method

Vexify4103 committed:
refactor: adjust token refresher timing for tighter expiry handling

Changed the anonymous token refresher to check every 1 second instead of every 10 seconds,
and trigger refresh 100ms after expiry instead of 1 minute before.

This ensures the refresher reacts much closer to the actual expiration point,
avoiding long gaps where an expired token could remain unrefreshed for several seconds.
The shorter ticker interval provides finer granularity for expiry detection and
reduces the risk of serving expired tokens between refresh cycles.

Vexify4103 committed:
feat: add initial fetch for anonymous Spotify token

Introduces logic to fetch and cache the initial anonymous Spotify token at the start of the token refresher.

Vexify4103 committed:
fix: json: invalid use of ,string struct tag

invalid JSON: json: invalid use of ,string struct tag, trying to unmarshal unquoted value into int64

topi314 commented on main.go:
```diff
@@ -74,11 +77,6 @@ func main() {
 	<-sig
 }
 
-type server struct {
-	ctx    context.Context
-	server *http.Server
-}
-
```
why was this moved?

topi314 commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/topi314/spotify-tokener
 
-go 1.24
+go 1.24.0
```
don't touch that, thanks

topi314 commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/topi314/spotify-tokener
```
you should run `go mod tidy`

topi314 commented on handler.go:
```diff
@@ -24,7 +46,43 @@ func (s *server) handleToken(w http.ResponseWriter, r *http.Request) {
+	// -------------------
+
+	// Use singleflight to prevent duplicate fetches
+	v, err, _ := s.group.Do(key, func() (interface{}, error) {
```
pretty sure you should call this before you do any processing for it to work correctly
we also don't use `interface{}` anymore

topi314 commented on handler.go:
```diff
@@ -100,3 +159,90 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+func (s *server) startAnonymousTokenRefresher() {
+	slog.Info("Fetching initial anonymous Spotify token...")
+
+	body, err := s.getAccessTokenPayload(s.ctx, nil)
+	if err != nil {
+		slog.Error("Failed to fetch initial anonymous token", slog.Any("err", err))
+	} else {
+		exp, err := parseExpiry(body)
+		if err != nil {
+			slog.Error("Failed to parse initial anonymous token expiry", slog.Any("err", err))
+		} else {
+			s.cacheMutex.Lock()
+			s.tokenCache[""] = cachedTokenEntry{
+				Token:  body,
+				Expiry: exp,
+			}
+			s.cacheMutex.Unlock()
+			slog.Info("Initial anonymous Spotify token fetched successfully", slog.Time("expiry", exp))
+		}
+	}
```
why does this duplicate code exist?

topi314 commented on handler.go:
```diff
@@ -100,3 +159,90 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+		}
+	}
+	
+	ticker := time.NewTicker(1 * time.Second)
```
you dont need to tick every second imo
just make it like once a minute...

topi314 commented on handler.go:
```diff
@@ -100,3 +159,90 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+	
+	for range ticker.C {
+		s.cacheMutex.Lock()
+		entry, exists := s.tokenCache[""]
```
this should basically always exist?
if not a new should be fetched anyway...

topi314 commented on handler.go:
```diff
@@ -100,3 +159,90 @@ func (s *server) getAccessTokenPayload(rCtx context.Context, cookies []*network.
+		}
+
+		exp, err := parseExpiry(body)
+
```
random empty line?
